### PR TITLE
Expose Ledger events through the API.

### PR DESCRIPTION
**Pull request by nc6**
- Add a function 'applyBlockWithEvents' which yields events along with a
  new ledger state.

### CODE REVIEW

File: cardano-api/src/Cardano/Api/LedgerState.hs [hunk 1]
```diff
@@ -881,7 +895,22 @@ applyBlock' env oldState enableValidation block = do
...
+  -- ^ True to validate
+  ->  HFC.HardForkBlock
+            (Consensus.CardanoEras Consensus.StandardCrypto)
+  -> Either Text LedgerStateEvents
+applyBlockWithEvents env oldState enableValidation block = do
+  let config = envLedgerConfig env
+      stateOld = clsState oldState
+  if enableValidation
+    then tickThenApply config block stateOld
+    else tickThenReapplyCheckHash config block stateOld
```
Comment (newhoggy) [inline cardano-api/src/Cardano/Api/LedgerState.hs:913]: New `applyBlockWithEvents` function.

File: cardano-api/src/Cardano/Api/LedgerState.hs [hunk 2]
```diff
@@ -928,13 +975,11 @@ tickThenApply
...
         (HFC.HardForkBlock
             (Consensus.CardanoEras Shelley.StandardCrypto))
-    -> Either Text (Shelley.LedgerState
-        (HFC.HardForkBlock
-            (Consensus.CardanoEras Shelley.StandardCrypto)))
+    -> Either Text LedgerStateEvents
 tickThenApply cfg block lsb
-  = either (Left . Text.pack . show) Right
-  $ runExcept
-  $ Ledger.tickThenApply cfg block lsb
```
Comment (nfrisby) [inline cardano-api/src/Cardano/Api/LedgerState.hs:937]: I just pushed a refined interface to PR input-output-hk/ouroboros-network#3292

I think this should now just be `... $ runExcept $ Ledger.tickThenApplyLedgerResult cfg block lsb`, no need for `runLedgerT`.

File: cardano-api/src/Cardano/Api/LedgerState.hs [hunk 3]
```diff
@@ -892,12 +940,11 @@ tickThenReapplyCheckHash
     -> Shelley.LedgerState
         (HFC.HardForkBlock
             (Consensus.CardanoEras Shelley.StandardCrypto))
-    -> Either Text (Shelley.LedgerState
-        (HFC.HardForkBlock
-            (Consensus.CardanoEras Shelley.StandardCrypto)))
+    -> Either Text LedgerStateEvents
 tickThenReapplyCheckHash cfg block lsb =
   if Consensus.blockPrevHash block == Ledger.ledgerTipHash lsb
-    then Right $ Ledger.tickThenReapply cfg block lsb
```
Comment (nfrisby) [inline cardano-api/src/Cardano/Api/LedgerState.hs:900]: I just pushed a refined interface to PR input-output-hk/ouroboros-network#3292

I think this should now just be `... $ Ledger.tickThenReapplyLedgerResult cfg block lsb`, no need for `runIdentity` or `runLedgerT`.

**Review by nfrisby — COMMENTED** [review]
I just pushed a refined interface to PR input-output-hk/ouroboros-network#3292 and left a couple comments here explaining what needs to change. HTH.

**Review by erikd — APPROVED** [review]

**Comment by nc6** [comment]
bors merge

**Pull request merged**